hedera-node/src/main/java/com/hedera/services/stream/RecordStreamFileWriter.java

```diff
@@ -0,0 +1,490 @@
+	 * @param consensusTimestamp
+	 * 	the consensus timestamp of the first transaction in the record file
+	 * @return
+	 * 	the new record file name
+	 */
+	String generateStreamFilePath(final Instant consensusTimestamp) {
+		return dirPath + File.separator + generateStreamFileNameFromInstant(consensusTimestamp, streamType);
+	}
+
+	public void setRunningHash(final Hash hash) {
+		this.runningHash = new RunningHash(hash);
+	}
+
+	/**
+	 * this method is called when the node falls behind
+	 * resets all populated up to this point fields (metadataDigest, dosMeta, recordStreamFile)
+	 */
+	@Override
+	public void clear() {
+		if (dosMeta != null) {
+			try {
+				dosMeta.close();
+				metadataStreamDigest.reset();
+				dosMeta = null;
+			} catch (IOException e) {
+				LOG.warn(EXCEPTION.getMarker(), "RecordStreamFileWriter::clear Exception in closing dosMeta", e);
+			}
+		}
+		recordStreamFileBuilder = null;
+		LOG.debug(OBJECT_STREAM.getMarker(), "RecordStreamFileWriter::clear executed.");
+	}
+
+	public void close() {
+		this.closeCurrentAndSign();
+		LOG.debug(LogMarker.FREEZE.getMarker(), "RecordStreamFileWriter finished writing the last object, is stopped");
+	}
+
+	public void setStartWriteAtCompleteWindow(final boolean startWriteAtCompleteWindow) {
+		this.startWriteAtCompleteWindow = startWriteAtCompleteWindow;
+		LOG.debug(OBJECT_STREAM.getMarker(),
+				"RecordStreamFileWriter::setStartWriteAtCompleteWindow: {}", startWriteAtCompleteWindow);
+	}
+
+	public boolean getStartWriteAtCompleteWindow() {
+		return this.startWriteAtCompleteWindow;
+	}
+
+	/**
+	 * Helper method that serializes a RecordStreamFile.
+	 * Uses deterministic serialization to ensure multiple invocations on the same
+	 * object lead to identical serialization.
+	 * @param recordStreamFileProtoBuilder
+	 * 		the RecordStreamFile object that needs to be serialized
+	 * @return
+	 * 		the serialized bytes
+	 */
+	private byte[] serialize(final RecordStreamFile.Builder recordStreamFileProtoBuilder) throws IOException {
+		final var recordStreamFileProto = recordStreamFileProtoBuilder.build();
+		final var result = new byte[recordStreamFileProto.getSerializedSize()];
+		final var output = CodedOutputStream.newInstance(result);
+		output.useDeterministicSerialization();
+		recordStreamFileProto.writeTo(output);
+		output.checkNoSpaceLeft();
+		return result;
+	}
+
+	private HashObject toProto(final Hash hash) {
+		return HashObject.newBuilder()
+				.setAlgorithm(HashAlgorithm.SHA_384)
+				.setLength(hash.getDigestType().digestLength())
+				.setHash(ByteString.copyFrom(hash.getValue()))
+				.build();
+	}
+
+	private void createSignatureFile(final File relatedRecordStreamFile) {
+		// get entire hash of current record stream file
+		final var entireHash = new Hash(streamDigest.digest(), DigestType.SHA_384);
+		// get metaData hash of current record stream file
+		final var metaHash = new Hash(metadataStreamDigest.digest(), DigestType.SHA_384);
```

Comment:
Why use a temporary `Hash` object as a container just to convert it to proto `HashObject` right after? Should just use proto directly.

Reply:
[Fixed](https://github.com/hashgraph/hedera-services/pull/3461/commits/4cc3f212e7d35ef866331a8df180ba914b6d3ca7).